test/hotspot/jtreg/compiler/lib/generators/Generators.java

```diff
@@ -0,0 +1,573 @@
+    }
+
+    /**
+     * Returns a new restrictable generator that samples its next element from either generator A or B, with assignable weights.
+     * Restricting this generator restricts each subgenerator. Generators which become empty by the restriction are
+     * removed from the new mixed generator. Weights stay their original value if a generator is removed. If the mixed
+     * generator would become empty by applying a restriction {@link EmptyGeneratorException} is thrown.
+     */
+    public <T extends Comparable<T>> RestrictableGenerator<T> mixed(RestrictableGenerator<T> a, RestrictableGenerator<T> b, int weightA, int weightB) {
+        return new RestrictableMixedGenerator<>(this, List.of(a, b), List.of(weightA, weightB));
+    }
+
+    /**
+     * Returns a new restrictable generator that samples its next element randomly from one of the provided restrictable
+     * generators with assignable weights.
+     * See {@link #mixed(RestrictableGenerator, RestrictableGenerator, int, int)} for details about restricting this
+     * generator.
+     */
+    @SafeVarargs
+    public final <T extends Comparable<T>> RestrictableGenerator<T> mixed(List<Integer> weights, RestrictableGenerator<T>... generators) {
+        return new RestrictableMixedGenerator<>(this, Arrays.asList(generators), weights);
+    }
+
+    /**
+     * Randomly pick an int generator.
+     *
+     * @return Random int generator.
+     */
+    public RestrictableGenerator<Integer> ints() {
+        switch(random.nextInt(0, 6)) {
+            case 0  -> { return uniformInts(); }
+            case 1  -> { return specialInts(0); }
+            case 2  -> { return specialInts(2); }
+            case 3  -> { return specialInts(16); }
+            case 4  -> { return uniformIntsMixedWithSpecials(1, 1, 16); }
+            case 5  -> { return uniformIntsMixedWithSpecials(1, 2, 2); }
+            default -> { throw new RuntimeException("impossible"); }
+        }
+    }
+
+    /**
+     * A generator of special ints. Special ints are powers of two or values close to powers of 2, where a value
+     * is close to a power of two p if it is in the interval [p - range, p + range]. Note that we also consider negative
+     * values as powers of two.
+     */
+    public RestrictableGenerator<Integer> specialInts(int range) {
+        TreeSet<Integer> set = new TreeSet<>();
+        for (int i = 0; i < 32; i++) {
+            int pow2 = 1 << i;
+            for (int j = -range; j <= range; j++) {
+                set.add(+pow2 + j);
+                set.add(-pow2 + j);
+            }
+        }
+        return orderedRandomElement(set);
+    }
+
+    /**
+     * A convenience helper to mix {@link #specialInts(int)} with {@link #uniformInts(int, int)}.
+     */
+    public RestrictableGenerator<Integer> uniformIntsMixedWithSpecials(int weightA, int weightB, int rangeSpecial) {
+        return mixed(uniformInts(), specialInts(rangeSpecial), weightA, weightB);
+    }
```

Comment:
I think you should give the weights a better name -> weightUniform and weightSpecial ?